### ISSUE TITLE
Wrong typing after yield

### ISSUE THREAD

**Issue by juk-r** [microsoft/pyright#6254]
Wrong determines type (value) of variable after yield expression, because it could have changed.
```python
class A:
    def __init__(self):
        self.x: int = 0
        self.y: int | str = 0

    def fun(self):
        self.x = 0
        self.y = 0
        yield None
        if self.x == 1: # reportUnnecessaryComparison
            pass
        if isinstance(self.y, int): # reportUnnecessaryComparison
            pass
```

**Comment by erictraut** [comment]
This behavior is intended. It's not specific to `yield`. The same thing occurs for any function call.

```python
class A:
    def __init__(self):
        self.y: int | str = 0

    def fun(self):
        self.y = 0
        reveal_type(self.y)  # Literal[0]
        self.modify_y()
        reveal_type(self.y)  # Literal[0]

    def modify_y(self):
        self.y = ""
```

When type checkers like pyright or mypy narrow the type of an attribute based on an assignment, they assume that the attribute will be subsequently modified only using local side effects (those that are observable within that function). Your code has non-local side effects that affect the value of `x` and `y`. If a type checker did not make this assumption, type narrowing for attribute access expressions would be useless in general.
